src/Build/Logging/ReusableLogger.cs

```diff
@@ -0,0 +1,585 @@
+    {
+        // Create the handlers.
+        _anyEventHandler = AnyEventRaisedHandler;
+        _buildFinishedEventHandler = BuildFinishedHandler;
+        _buildStartedEventHandler = BuildStartedHandler;
+        _customBuildEventHandler = CustomEventRaisedHandler;
+        _buildErrorEventHandler = ErrorRaisedHandler;
+        _buildMessageEventHandler = MessageRaisedHandler;
+        _projectFinishedEventHandler = ProjectFinishedHandler;
+        _projectStartedEventHandler = ProjectStartedHandler;
+        _buildStatusEventHandler = StatusEventRaisedHandler;
+        _targetFinishedEventHandler = TargetFinishedHandler;
+        _targetStartedEventHandler = TargetStartedHandler;
+        _taskFinishedEventHandler = TaskFinishedHandler;
+        _taskStartedEventHandler = TaskStartedHandler;
+        _buildWarningEventHandler = WarningRaisedHandler;
+        _telemetryEventHandler = TelemetryLoggedHandler;
+        _workerNodeTelemetryLoggedHandler = WorkerNodeTelemetryLoggedHandler;
+
+        // Register for the events.
+        eventSource.AnyEventRaised += _anyEventHandler;
+        eventSource.BuildFinished += _buildFinishedEventHandler;
+        eventSource.BuildStarted += _buildStartedEventHandler;
+        eventSource.CustomEventRaised += _customBuildEventHandler;
+        eventSource.ErrorRaised += _buildErrorEventHandler;
+        eventSource.MessageRaised += _buildMessageEventHandler;
+        eventSource.ProjectFinished += _projectFinishedEventHandler;
+        eventSource.ProjectStarted += _projectStartedEventHandler;
+        eventSource.StatusEventRaised += _buildStatusEventHandler;
+        eventSource.TargetFinished += _targetFinishedEventHandler;
+        eventSource.TargetStarted += _targetStartedEventHandler;
+        eventSource.TaskFinished += _taskFinishedEventHandler;
+        eventSource.TaskStarted += _taskStartedEventHandler;
+        eventSource.WarningRaised += _buildWarningEventHandler;
+
+        if (eventSource is IEventSource2 eventSource2)
+        {
+            eventSource2.TelemetryLogged += _telemetryEventHandler;
+        }
+
+        if (eventSource is IEventSource3 eventSource3)
+        {
+            if (_includeEvaluationMetaprojects)
+            {
+                eventSource3.IncludeEvaluationMetaprojects();
+            }
+
+            if (_includeEvaluationProfiles)
+            {
+                eventSource3.IncludeEvaluationProfiles();
+            }
+
+            if (_includeTaskInputs)
+            {
+                eventSource3.IncludeTaskInputs();
+            }
+        }
+
+        if (eventSource is IEventSource4 eventSource4)
+        {
+            if (_includeEvaluationPropertiesAndItems)
+            {
+                eventSource4.IncludeEvaluationPropertiesAndItems();
+            }
+        }
+
+        if (eventSource is IEventSource5 eventSource5)
+        {
+            eventSource5.WorkerNodeTelemetryLogged += _workerNodeTelemetryLoggedHandler;
+        }
```

Comment:
Added IEventSource5 support to this in anticipation of it going public at some point.